### PR TITLE
DS2-154: History document syncing overwrites document of deleted entity when a company with same final_url is created

### PR DESCRIPTION
**Pull request by GuillermoDR**
Fixed bug.
If we want to add a company with the same final_url as another that was previously deleted. We need to check that the dealroom_id is different and add a NEW document.

### CODE REVIEW

**Review by Meenu-Jain — APPROVED** [review]

File: dealroom_firestore_connector/__init__.py [hunk 1]
```diff
@@ -407,6 +407,31 @@ def _get_final_url_and_dealroom_id(
     return final_url, dealroom_id
 
 
+def check_for_deleted_profiles(
+    key_found, history_refs, dealroom_id, count_history_refs
+):
+    # https://dealroom.atlassian.net/browse/DS2-154
+    document_matches_by_final_url = key_found == "final_url"
```
Comment (vittorio-dealroom) [inline dealroom_firestore_connector/__init__.py:414]: I would put this check outside of this function: would you evaluate the condition to call a function outside the function, or would you evaluate if the function needs to do smthg after you've called it?

**Review by vittorio-dealroom — CHANGES_REQUESTED** [review]

File: dealroom_firestore_connector/__init__.py (hunk 1 above)
Comment (GuillermoDR) [reply dealroom_firestore_connector/__init__.py:414]: Evaluating a condition is less complex since it's just one line in machine code, while calling a function requires at least two. So I'll update it

Comment (GuillermoDR) [reply dealroom_firestore_connector/__init__.py]: Updated in [7a914f9](https://github.com/dealroom/data-firestore-connector/pull/28/commits/7a914f91e82b68cb2a41ee63d610d3b9893b00a9)

Comment (vittorio-dealroom) [reply dealroom_firestore_connector/__init__.py]: ok, but really what's important here is readability: evaluating the condition inside the function forces the reader to read code that is nested more than it should be. Evaluating the condition outside the function also improves readability, since you already are calling the funct conditionally

File: dealroom_firestore_connector/__init__.py [hunk 2]
```diff
@@ -407,6 +407,29 @@ def _get_final_url_and_dealroom_id(
     return final_url, dealroom_id
 
 
+def check_for_deleted_profiles(
+    key_found, history_refs, dealroom_id, count_history_refs
```
Comment (vittorio-dealroom) [inline dealroom_firestore_connector/__init__.py:411]:
```suggestion
    doc_refs, dealroom_id, count_history_refs
```
you do not need `key_found` as arg, since the condition to call this function is already checking that that variable has the correct value. what you need here as arg is `doc_refs`, which when you call the function is `doc_refs = history_refs[key_found]`
@GuillermoDR

File: dealroom_firestore_connector/__init__.py [hunk 3]
```diff
@@ -458,6 +482,12 @@ def set_history_doc_refs(
     else:
         count_history_refs = 0
 
+    document_matches_by_final_url = key_found == "final_url"
+    if document_matches_by_final_url:
```
Comment (vittorio-dealroom) [inline dealroom_firestore_connector/__init__.py:486]: why was this removed? `if count_history_refs:` @GuillermoDR

Comment (GuillermoDR) [reply dealroom_firestore_connector/__init__.py:486]: It's the same thing, `count_history_refs` will be initialised if `document_matches_by_final_url` has a value

File: dealroom_firestore_connector/__init__.py (hunk 2 above)
Comment (GuillermoDR) [reply dealroom_firestore_connector/__init__.py]: Fixed in [452d4a4](https://github.com/dealroom/data-firestore-connector/pull/28/commits/452d4a4abaa4da12b01b4e7f75877ce1ddf6d3af)

File: dealroom_firestore_connector/__init__.py [hunk 4]
```diff
@@ -407,6 +407,29 @@ def _get_final_url_and_dealroom_id(
...
+    doc_refs, dealroom_id, count_history_refs
+):
+    # https://dealroom.atlassian.net/browse/DS2-154
+    # check all matching docs
+    for doc_ref in doc_refs:
+        doc = doc_ref.get().to_dict()
+        is_a_deleted_entity = doc["dealroom_id"] == _DELETED_DEALROOM_ENTITY_ID
+        dealroom_id_was_already_used = (
+            "dealroom_id_old" in doc and doc["dealroom_id_old"] == dealroom_id
+        )
```
Comment (vittorio-dealroom) [inline dealroom_firestore_connector/__init__.py:420]:
```suggestion
        dealroom_id_was_already_used = doc.get("dealroom_id_old") == dealroom_id
```

**Review by vittorio-dealroom — APPROVED** [review]

**Review by vittorio-dealroom — APPROVED** [review]
